### PR TITLE
[WIP] :bug: Support chunking AWS secrets

### LINKED ISSUE #1531: Validation exception because secret crosses 10k limit on creation of a machine
/kind bug
I am seeing validation exception because secret crosses 10k limit on creation of a machine.
"msg"="Reconciler error" "error"="ValidationException: 1 validation error detected: Value at 'secretBinary' failed to satisfy constraint: Member must have length less than or equal to 10240\n\tstatus code: 400

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
Used release 0.4 staging image to create a cluster
Probably we should add it into multiple secrets and bind them together.

**Environment:**

- Cluster-api-provider-aws version: 0.4 staging -> 0.4.9
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by ncdc**
**What this PR does / why we need it**:
AWS has a 10kb limit on binary secret data (base64 encoded). This chunks
secrets to stay under the 10kb limit and stitches them back together in
the VM during cloud-init.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1531

### CODE REVIEW

File: pkg/cloud/services/secretsmanager/cloudinit.go [hunk 1]
```diff
@@ -65,8 +66,9 @@ func GenerateCloudInitMIMEDocument(secretID, region string) ([]byte, error) {
 	}
 
 	scriptVariables := scriptVariables{
-		SecretID: secretID,
-		Region:   region,
+		SecretPrefix: secretPrefix,
+		Chunks:       chunks - 1,
```
Comment (randomvariable) [inline pkg/cloud/services/secretsmanager/cloudinit.go]: this is a cheat for avoiding bash, shouldn't do this.

Comment (ncdc) [reply pkg/cloud/services/secretsmanager/cloudinit.go]: undo, then in bash, we can do `CHUNK_SEQUENCE_END=$((CHUNKS-1))`

File: api/v1alpha2/awsmachine_conversion_test.go [hunk 2]
```diff
@@ -34,15 +34,15 @@ func TestConvertAWSMachine(t *testing.T) {
 				Spec: infrav1alpha3.AWSMachineSpec{
 					CloudInit: infrav1alpha3.CloudInit{
 						InsecureSkipSecretsManager: true,
-						SecretARN:                  "something",
+						SecretPrefix:               "something/",
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply api/v1alpha2/awsmachine_conversion_test.go]: yeah, need to change this. assume it comes from ncdc testing various schemes

File: pkg/cloud/services/secretsmanager/secret_fetch_script.go [hunk 3]
```diff
@@ -33,123 +33,163 @@ const secretFetchScript = `#!/bin/bash
...
+    log::error_exit "Could not delete secret value" 2
+  fi
+}
+
+delete_secrets() {
+  for i in $(seq 0 ${FINAL_INDEX}); do
+    delete_secret_value "$i"
+    true
+  done
+}
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply pkg/cloud/services/secretsmanager/secret_fetch_script.go]: we do both. controller will mop up if the machine fails.

File: pkg/cloud/services/secretsmanager/secret_fetch_script.go [hunk 4]
```diff
@@ -33,123 +33,163 @@ const secretFetchScript = `#!/bin/bash
...
+set -o nounset
+set -o pipefail
+
+umask 006
+
 REGION="{{.Region}}"
-SECRET_ID="{{.SecretID}}"
+SECRET_PREFIX="{{.SecretPrefix}}"
+CHUNKS="{{.Chunks}}"
 FILE="/etc/secret-userdata.txt"
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply pkg/cloud/services/secretsmanager/secret_fetch_script.go]: this is fairly low risk given `/etc/kubernetes/pki` is also pre-known. it is chmod'd to be root only. feel free to open a new issue if it's important.

**Comment by randomvariable** [comment]
Need to resubmit this to squash

File: pkg/cloud/services/secretsmanager/secret.go [hunk 5]
```diff
@@ -50,40 +58,56 @@ func (s *Service) Create(m *scope.MachineScope, data []byte) (string, error) {
...
-	_, err := s.scope.SecretsManager.DeleteSecret(&secretsmanager.DeleteSecretInput{
-		SecretId:                   aws.String(secretArn),
-		ForceDeleteWithoutRecovery: aws.Bool(true),
-	})
-
-	if awserrors.IsNotFound(err) {
-		return nil
+	var errors []error
+
+	for i := int32(0); i < m.GetSecretCount(); i++ {
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply pkg/cloud/services/secretsmanager/secret.go]: could abort too early if it was aborted prematurely (controller aborted or machine failed to delete its own)

**Comment by randomvariable** [comment]
superseded by
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1535

**Pull request closed without merging**